compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaComposition.cpp

```diff
@@ -38,27 +38,54 @@ class AMDAIEDmaCompositionPass
   Operation *parentOp = getOperation();
   MLIRContext *context = &getContext();
-  RewritePatternSet patterns(context);
+
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(parentOp);
+  std::optional<AMDAIEDevice> maybeDevice = getConfigAMDAIEDevice(targetAttr);
+  if (!maybeDevice) {
+    parentOp->emitOpError()
+        << "has no AMDAIEDevice in the target attribute configuration. This "
+           "device-specific information is required to determine when loops "
+           "can be subsumed into DMA operations, and must be attached to a "
+           "containing ModuleOp.";
+    return signalPassFailure();
+  }
+  AMDAIE::AMDAIEDeviceModel deviceModel =
+      AMDAIE::getDeviceModel(maybeDevice.value());
+
+  {
+    RewritePatternSet patterns(context);
+    populateDmaLoopSubsumptionPattern(patterns, std::move(deviceModel),
+                                      onlyZeroStrideOnOuterDim);
+    if (failed(applyPatternsAndFoldGreedily(parentOp, std::move(patterns)))) {
+      parentOp->emitOpError("failed to subsume loops into DMA operations");
+      return signalPassFailure();
+    }
+  }
   {
-    auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(parentOp);
-    std::optional<AMDAIEDevice> maybeDevice = getConfigAMDAIEDevice(targetAttr);
-    if (!maybeDevice) {
-      parentOp->emitOpError()
-          << "has no AMDAIEDevice in the target attribute configuration. This "
-             "device-specific information is required to determine when loops "
-             "can be subsumed into DMA operations, and must be attached to a "
-             "containing ModuleOp.";
+    RewritePatternSet patterns(context);
+    populateStridedOpCombinationPattern(patterns);
+    if (failed(applyPatternsAndFoldGreedily(parentOp, std::move(patterns)))) {
+      parentOp->emitOpError("failed to compose strided operations");
       return signalPassFailure();
     }
-    AMDAIE::AMDAIEDeviceModel deviceModel =
-        AMDAIE::getDeviceModel(maybeDevice.value());
+  }
+  {
+    RewritePatternSet patterns(context);
+    populateCanonicalizeDoublyStridedOpPatterns(patterns, false);
+    if (failed(applyPatternsAndFoldGreedily(parentOp, std::move(patterns)))) {
+      parentOp->emitOpError(
+          "failed to canonicalize doubly strided DMA operations");
+      return signalPassFailure();
+    }
+  }
+  {
+    RewritePatternSet patterns(context);
     populateDmaLoopSubsumptionPattern(patterns, std::move(deviceModel),
                                       onlyZeroStrideOnOuterDim);
-  }
-  populateStridedOpCombinationPattern(patterns);
-  if (failed(applyPatternsAndFoldGreedily(parentOp, std::move(patterns)))) {
-    parentOp->emitOpError("failed to compose strided operations");
-    return signalPassFailure();
+    if (failed(applyPatternsAndFoldGreedily(parentOp, std::move(patterns)))) {
+      parentOp->emitOpError("failed to compose strided operations");
+      return signalPassFailure();
+    }
```

Comment:
Applying the patterns one by one will always miss some composition opportunities. We should be populating all patterns and applying them together. This will be the most versatile and keep rewriting until all DMA composition opportunities are exhausted:

```
RewritePatternSet patterns(context);
populateDmaLoopSubsumptionPattern(patterns, std::move(deviceModel),
                                      onlyZeroStrideOnOuterDim);
populateStridedOpCombinationPattern(patterns);
populateCanonicalizeDoublyStridedOpPatterns(patterns, false);
```

Reply:
We've got a couple of passes which look like they're doing canonicalization
of DMA ops. Can they be put into a single AMDAIECanonicalizeDma.cpp, which 
we then just run like vanilla canonicalization? 

AMDAIECanonicalizeDoublyStridedOp.cpp
AMDAIECanonicalizeNpuDmaCpyNd.cpp
AMDAIECombineStridedOps.cpp
AMDAIEDmaComposition.cpp
AMDAIEDmaCSE.cpp
AMDAIEDmaLoopSubsumption.cpp
AMDAIENormalizeLoopBounds.cpp

Reply:
(obviously not for this PR, but maybe some refactoring for later)

Reply:
> Applying the patterns one by one will always miss some composition opportunities. We should be populating all patterns and applying them together. This will be the most versatile and keep rewriting until all DMA composition opportunities are exhausted:
> 
> ```
> RewritePatternSet patterns(context);
> populateDmaLoopSubsumptionPattern(patterns, std::move(deviceModel),
>                                       onlyZeroStrideOnOuterDim);
> populateStridedOpCombinationPattern(patterns);
> populateCanonicalizeDoublyStridedOpPatterns(patterns, false);
> ```

I still couldn't make it work as expected if populate all patterns and make them under the same RewritePatternSet. @jtuyls Could you try it and give some suggestions?

Reply:
> %48 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[0, 0, %45] [4, 32, 64] [4096, 128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>

Isn't this the same as:

```
%48 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[0, %45] [128, 64] [128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>
```

I think you can adjust `CombineStridedOps`, so the following is combined into the result above:

```
48 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[0, %45] [32, 64] [128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>
%49 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[32, %45] [96, 64] [128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>
```

Reply:
I meant canonicalization in a  general sense. Canonicalization in the strict sense should be done directly with op class methods. I don't want to spend time arguing abstract points for which I have no evidence but my intuition says it's good to have lots of _lowering_ passes, but few, general purpose canonicalization/cse/normalization passes which can be run anywhere.

Reply:
> I think you can adjust `CombineStridedOps`, so the following is combined into the result above:
> 
> ```
> 48 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[0, %45] [32, 64] [128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>
> %49 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[32, %45] [96, 64] [128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>
> ```

Are you sure this is combinable for a more general case regardless of what we already know from this example? It is not consistent with the current logic that inserting a new dimension if two npu has the same number of dimensions of addressing.

Reply:
> > I think you can adjust `CombineStridedOps`, so the following is combined into the result above:
> > ```
> > 48 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[0, %45] [32, 64] [128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>
> > %49 = amdaie.npu.dma_cpy_nd %8([] [] [], %31[32, %45] [96, 64] [128, 1]) : source_type = !amdaie.logicalobjectfifo<memref<16384xi32>>
> > ```
> 
> Are you sure this is combinable for a more general case regardless of what we already know from this example?

Yes, because the above two DMA ops are just an extension of each other, so no new dimension is needed. This is an edge case which is just not handled currently.

Reply:
> I meant canonicalization in a general sense. Canonicalization in the strict sense should be done directly with op class methods.

In that sense, you could argue that every lowering pass is a performing a canonicalization. so that the `canonical` form is the most efficiently lowered form for the given hardware, and therefore you could put the whole codegen inside a single canonicalization pass. That's against the point of having lowering passes, so I think we shouldn't look at it like that, and unless we have a very good reason to deviate, we should stay with the definition and use of canonicalization within MLIR: https://mlir.llvm.org/docs/Canonicalization/.